crates/iceberg/src/expr/visitors/expression_evaluator.rs

```diff
@@ -0,0 +1,819 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use fnv::FnvHashSet;
+
+use crate::{
+    expr::{BoundPredicate, BoundReference},
+    spec::{DataFile, Datum, PrimitiveLiteral, Struct},
+    Error, ErrorKind, Result,
+};
+
+use super::bound_predicate_visitor::{visit, BoundPredicateVisitor};
+
+/// Evaluates a [`DataFile`]'s partition [`Struct`] to check
+/// if the partition tuples match the given [`BoundPredicate`].
+///
+/// Use within [`TableScan`] to prune the list of [`DataFile`]s
+/// that could potentially match the TableScan's filter.
+#[derive(Debug)]
+pub(crate) struct ExpressionEvaluator {
+    /// The provided partition filter.
+    partition_filter: BoundPredicate,
+}
+
+impl ExpressionEvaluator {
+    /// Creates a new [`ExpressionEvaluator`].
+    pub(crate) fn new(partition_filter: BoundPredicate) -> Self {
+        Self { partition_filter }
+    }
+
+    /// Evaluate this [`ExpressionEvaluator`]'s partition filter against
+    /// the provided [`DataFile`]'s partition [`Struct`]. Used by [`TableScan`]
+    /// to see if this [`DataFile`] could possibly contain data that matches
+    /// the scan's filter.
+    pub(crate) fn eval(&self, data_file: &DataFile) -> Result<bool> {
+        let mut visitor = ExpressionEvaluatorVisitor::new(self, data_file.partition());
+
+        visit(&mut visitor, &self.partition_filter)
+    }
+}
+
+/// Acts as a visitor for [`ExpressionEvaluator`] to apply
+/// evaluation logic to different parts of a data structure,
+/// specifically for data file partitions.
+#[derive(Debug)]
+struct ExpressionEvaluatorVisitor<'a> {
+    /// Reference to an [`ExpressionEvaluator`].
+    expression_evaluator: &'a ExpressionEvaluator,
+    /// Reference to a [`DataFile`]'s partition [`Struct`].
+    partition: &'a Struct,
+}
+
+impl<'a> ExpressionEvaluatorVisitor<'a> {
+    /// Creates a new [`ExpressionEvaluatorVisitor`].
+    fn new(expression_evaluator: &'a ExpressionEvaluator, partition: &'a Struct) -> Self {
+        Self {
+            expression_evaluator,
+            partition,
+        }
+    }
+
+    /// Checks if the [`PrimitiveLiteral`] is null.
+    fn is_null(literal: &PrimitiveLiteral) -> bool {
+        if let PrimitiveLiteral::Boolean(false) = literal {
```

Comment:
when the `DataFile` contains a partition struct like:
```Rust
let partition = Struct::from_iter([None]);
```

the `Struct` fields will contain a `Literal::Primitive(PrimitiveLiteral::Boolean(false))`

```Rust
impl FromIterator<Option<Literal>> for Struct {
    fn from_iter<I: IntoIterator<Item = Option<Literal>>>(iter: I) -> Self {
        let mut fields = Vec::new();
        let mut null_bitmap = BitVec::new();

        for value in iter.into_iter() {
            match value {
                Some(value) => {
                    fields.push(value);
                    null_bitmap.push(false)
                }
                None => {
                    fields.push(Literal::Primitive(PrimitiveLiteral::Boolean(false)));
                    null_bitmap.push(true)
                }
            }
        }
        Struct {
            fields,
            null_bitmap,
        }
    }
}
```

thats why I check the `Result<Datum>` returned by the `Accessor`

Reply:
> Otherwise how do you distinguish it from actual value of `PrimitiveLiteral::Boolean(false)`. And the `Accessor` should return `Option<Literal>`.

Yeah, this makes sense to me. 
Regarding the `Accessor` will you open another issue or @sdd something you might want to take a look at?

Reply:
Hey @marvinlanhenke - I'll fix that `Accessor` bug, no problem.